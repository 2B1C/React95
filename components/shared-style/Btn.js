import styled from 'styled-components';

export default styled.button`
  background-color: #c3c7cb;
  padding: 5px 20px 6px;
  border: none;

  box-shadow: inset 1px 1px 0px 1px #ffffff, inset 0 0 0 1px #868a8e,
    1px 1px 0 1px #000;

  &:disabled {
    color: #868a8e;
  }

  &:focus {
    outline: none;
    box-shadow: inset 1px 1px 0px 1px #ffffff, inset -1px -1px 0px 1px #868a8e,
      1px 1px 0 2px #000;
  }

  &:active {
    padding: 6px 20px 5px;

    outline: 1px dotted #000;
    outline-offset: -5px;
    outline-width: 2px;

    box-shadow: inset 0 0 0 1px #868a8e, 0 0 0 1px #000;
  }
`;
