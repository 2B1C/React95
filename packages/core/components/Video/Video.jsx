import React, { useRef, useState } from 'react';
import styled from '@xstyled/styled-components';
import { th } from '@xstyled/system';

import Btn from '../shared-style/Btn';
import { Frame, Range, Icon } from '..';
import { useEffect } from 'react';

const VideoTag = styled.video`
  width: 100%;
`;

const Source = ({ src }) => (
  <source src={src} type={`video/${src.substring(src.length - 3)}`} />
);

const ControlBtn = styled(Btn)`
  display: inline-flex;
  justify-content: center;
  align-items: center;

  &&,
  &:active,
  &:focus {
    width: 25px;
    height: 25px;
    padding: 7;
  }
`;

const TitleBar = styled.div`
  height: 18px;
  margin-bottom: 2;

  color: ${th('colors.white')};
  padding: 2 2 0;

  display: flex;
  background-color: primary;
`;

const Controls = styled.div`
  display: flex;
  align-items: center;
  padding: 2;
`;

const arrayFy = (str) => [].concat(str);

function updateProgressBar(player, updateProgress) {
  const percentage = Math.floor((100 / player.duration) * player.currentTime);

  updateProgress(percentage);
}

const Video = ({ src, ...props }) => {
  const [playing, setPlaying] = useState(false);
  const [loadeddata, setLoadeddata] = useState(false);
  const [progress, setProgress] = useState(0);
  const player = useRef(null);
  const progressRef = useRef(null);

  const paths = arrayFy(src);
  const [name] = paths;

  useEffect(() => {
    player.current.addEventListener(
      'ended',
      () => {
        setPlaying(false);
        setProgress(0);
      },
      false,
    );
    player.current.addEventListener(
      'timeupdate',
      () => {
        updateProgressBar(player.current, setProgress);
      },
      false,
    );
    player.current.addEventListener(
      'loadeddata',
      () => {
        setLoadeddata(true);
      },
      false,
    );
  }, [player.current]);

  return (
    <Frame p={2}>
      <TitleBar>{name.replace(/^.*[\\\/]/, '')}</TitleBar>
      <VideoTag {...props} ref={player}>
        {paths.map((s) => (
          <Source src={s} />
        ))}
      </VideoTag>
      <Controls>
        <ControlBtn
          onClick={() => {
            if (!playing) {
              player.current.play();
            } else {
              player.current.pause();
            }
            setPlaying((playing) => !playing);
          }}
        >
          {loadeddata ? (playing ? 'Pause' : '►') : 'Waiting'}
        </ControlBtn>
        <ControlBtn
          onClick={() => {
            player.current.pause();
            if (player.current.currentTime) {
              player.current.currentTime = 0;
            }
            setPlaying(false);
          }}
        >
          ■
        </ControlBtn>

        <Range
          ref={progressRef}
          min="0"
          max="100"
          step="1"
          value={progress}
          onClick={(e) => {
            const { current: el } = progressRef;
            const { current: video } = player;

            const percent = e.nativeEvent.offsetX / el.offsetWidth;

            video.currentTime = percent * video.duration;

            setProgress(Math.floor(percent / 100));
          }}
        />
      </Controls>
    </Frame>
  );
};
export default Video;
