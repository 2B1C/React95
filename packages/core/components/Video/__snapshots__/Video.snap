// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Video /> Snapshot should match snapshot 1`] = `
.c12 {
  background-color: #c3c7cb;
  padding: 7px 20px 5px;
  border: none;
  color: #000000;
  font-size: 12px;
  box-shadow: inset 1px 1px 0px 1px #ffffff, inset 0 0 0 1px #868a8e, 1px 1px 0 0px #000000;
}

.c12:disabled {
  color: #868a8e;
}

.c12:focus {
  outline: 1px dotted #000000;
  outline-offset: -5px;
  box-shadow: inset 1px 1px 0px 1px #ffffff, inset -0.5px -0.5px 0px 1px #868a8e, 1px 1px 0 1px #000000;
}

.c12:active {
  padding: 8px 20px 4px;
  outline: 1px dotted #000000;
  outline-offset: -5px;
  box-shadow: inset 0 0 0 1px #868a8e, 0 0 0 1px #000000;
}

.c0 {
  background-color: #c3c7cb;
  width: auto;
  height: auto;
  box-shadow: inset 0.5px 0.5px 0px 0.5px #ffffff, inset 0 0 0 1px #868a8e, 1px 0px 0 0px #000000, 0px 1px 0 0px #000000, 1px 1px 0 0px #000000;
  padding: 2px;
}

.c1 {
  background-color: #c3c7cb;
  color: #fefefe;
  width: auto;
  height: auto;
  box-shadow: inset 0.5px 0.5px 0px 0.5px #ffffff, inset 0 0 0 1px #868a8e, 1px 0px 0 0px #000000, 0px 1px 0 0px #000000, 1px 1px 0 0px #000000;
}

.c5 {
  background-color: #c3c7cb;
  max-width: 250px;
  width: auto;
  height: auto;
  box-shadow: inset 0.5px 0.5px 0px 0.5px #ffffff, inset 0 0 0 1px #868a8e, 1px 0px 0 0px #000000, 0px 1px 0 0px #000000, 1px 1px 0 0px #000000;
  margin-right: auto;
  margin-left: auto;
  margin-bottom: 4px;
}

.c7 {
  background-color: #c3c7cb;
  width: auto;
  height: auto;
  box-shadow: inset 0.5px 0.5px 0px 0.5px #ffffff, inset 0 0 0 1px #868a8e, 1px 0px 0 0px #000000, 0px 1px 0 0px #000000, 1px 1px 0 0px #000000;
}

.c9 {
  background-color: #c3c7cb;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 40%;
  height: auto;
  box-shadow: inset 0.5px 0.5px 0px 0.5px #ffffff, inset 0 0 0 1px #868a8e, 1px 0px 0 0px #000000, 0px 1px 0 0px #000000, 1px 1px 0 0px #000000;
}

.c2 {
  height: 18px;
  margin-bottom: 2px;
  padding: 2px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-shadow: none;
  background: #000e7a;
}

.c2 img {
  width: 15px;
  height: 14px;
  margin-right: 4px;
}

.c3 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  font-weight: bold;
  line-height: 1.4em;
  margin: 0;
  font-size: 1em;
}

.c14 {
  -webkit-appearance: none;
  width: 100%;
}

.c14:hover,
.c14:focus,
.c14:active {
  outline: none;
}

.c14::-webkit-slider-runnable-track {
  width: 100%;
  height: 4px;
  background: #000000;
  cursor: pointer;
  border-top: 1px solid;
  border-top-color: #d2d2d2;
  border-right: 2px solid;
  border-right-color: #ffffff;
  border-bottom: 1px solid;
  border-bottom-color: #e6e6e6;
  border-left: 2px solid;
  border-left-color: #ffffff;
  box-shadow: 0px 1px 0px 0px #ffffff;
}

.c14::-webkit-slider-thumb {
  width: 12px;
  height: 20px;
  background-color: #c3c7cb;
  cursor: pointer;
  margin-top: -8px;
  border-top: 1px solid;
  border-top-color: #ffffff;
  border-right: 1px solid;
  border-right-color: #000000;
  border-bottom: 1px solid;
  border-bottom-color: #000000;
  border-left: 1px solid;
  border-left-color: #ffffff;
  box-shadow: inset 0px -1px 0px #868a8e, inset -1px 0px 0px #868a8e, inset 0px 1px 0px #e6e6e6, inset 1px 0px 0px #e6e6e6;
  -webkit-appearance: none;
}

.c14:focus::-webkit-slider-runnable-track {
  background: #000000;
}

.c14::-moz-range-track {
  width: 100%;
  height: 4px;
  background: #000000;
  cursor: pointer;
  border-top: 1px solid;
  border-top-color: #d2d2d2;
  border-right: 2px solid;
  border-right-color: #ffffff;
  border-bottom: 1px solid;
  border-bottom-color: #e6e6e6;
  border-left: 2px solid;
  border-left-color: #ffffff;
  box-shadow: 0px 1px 0px 0px #ffffff;
}

.c14::-moz-range-thumb {
  width: 12px;
  height: 20px;
  background-color: #c3c7cb;
  cursor: pointer;
  margin-top: -8px;
  border-top: 1px solid;
  border-top-color: #ffffff;
  border-right: 1px solid;
  border-right-color: #000000;
  border-bottom: 1px solid;
  border-bottom-color: #000000;
  border-left: 1px solid;
  border-left-color: #ffffff;
  box-shadow: inset 0px -1px 0px #868a8e, inset -1px 0px 0px #868a8e, inset 0px 1px 0px #e6e6e6, inset 1px 0px 0px #e6e6e6;
  -webkit-appearance: none;
}

.c4 {
  width: 100%;
  padding: 2px;
  display: none;
}

.c13 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c13 svg {
  fill: #000000;
}

.c13.c13,
.c13:active,
.c13:focus {
  width: 20px;
  height: 20px;
  padding: 7px;
  padding: 4px;
}

.c13.c13 svg,
.c13:active svg,
.c13:focus svg {
  fill: #868a8e;
  border-bottom: 1px solid;
  border-bottom-color: #ffffff;
  border-right: 1px solid;
  border-right-color: #ffffff;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 2px 0;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 6px;
  margin-bottom: 4px;
  box-shadow: inset 0px 0px 0px 0px, inset 1px 1px 0px 0px #868a8e, 0.5px 0.5px 0px 0.5px #ffffff;
  background-color: #000000;
  height: 50px;
  color: #ffffff;
}

.c10 {
  font-family: 'React95Video-Numbers';
  text-transform: uppercase;
}

.c6 {
  background-color: transparent;
  box-shadow: none;
}

.c15::-webkit-slider-thumb {
  height: 18px;
  margin-top: -7px;
  width: 10px;
}

.c15[value='0']::-webkit-slider-thumb {
  margin-left: -2px;
}

<div>
  <div
    class="c0"
    height="auto"
    style="width: 260px;"
    width="auto"
  >
    <div
      class="c1 c2"
      color="headerText"
      height="auto"
      width="auto"
    >
      <img
        height="16"
        src="test-file-stub"
        width="16"
      />
      <div
        class="c3"
      >
        some_video.mp4 (Opening)
      </div>
    </div>
    <video
      class="c4"
    >
      <source
        src="foo/bar/some_video.mp4"
        type="video/mp4"
      />
    </video>
    <div
      class="c5 c6"
      height="auto"
      width="auto"
    >
      <div
        class="c7 c8"
        height="auto"
        width="auto"
      >
        <div
          class="c9 c6"
          display="flex"
          height="auto"
          width="40%"
        >
          <span
            class="c10"
            style="margin-top: auto;"
          />
          <span
            class="c10"
            style="height: 12px;"
          >
            Openning
          </span>
        </div>
        <div
          class="c9 c6"
          display="flex"
          height="auto"
          width="40%"
        >
          <span
            class="c10"
            style="margin-top: auto; font-size: 22px;"
          />
          <span
            class="c10"
            style="height: 12px;"
          >
            time
          </span>
        </div>
      </div>
      <div
        class="c11"
      >
        <button
          class="c12 c13"
          disabled=""
        >
          <img
            height="32"
            src="test-file-stub"
            width="32"
          />
        </button>
        <button
          class="c12 c13"
          disabled=""
        >
          <svg
            aria-label="stop"
            height="6"
            width="6"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h306v306H0z"
            />
          </svg>
        </button>
        <input
          class="c14 c15"
          max="100"
          min="0"
          step="1"
          style="width: 70%; margin-left: 20px;"
          type="range"
          value="0"
        />
      </div>
    </div>
  </div>
</div>
`;
